### ISSUE TITLE
[dynamo] BUILD_STRING + SymNode in transformers' torch_compilable_check closure (19 VL models)

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#92]
**Repro status:** Reproduces on torch 2.12.0.dev20260407+cu128 (nightly, verified 2026-05-09 16:13 UTC). On torch 2.9.0+cu128 (stable, verified 2026-05-09 16:16 UTC), the same MRE triggers DIFFERENT graph breaks (gb0059 builtin `callable` + gb0055 NestedUserFunctionVariable proxy) — the BUILD_STRING path is specific to torch ≥2.10 (or thereabouts). Filing because the BUILD_STRING break is the dominant pattern observed across 19 VL models in the current sweep.

## Original failure report

19 vision-language / multimodal models produce 38 graph breaks from a single pattern: f-string assertions that include a SymNodeVariable in the formatted message. Dynamo hits gb0021 (`BUILD_STRING type error`) when the format string contains a SymNode part inside a closure-wrapped check.

**Test command (sweep):** `python tools/run_experiment.py sweep ...` (cohort: 19 VL models in eval+train, animesh-fullgraph 2026-04-28).

**Symptom captured:** every break's reason starts with `Graph break in user code at <transformers/models/.../modeling_*.py>:<line>` followed by `Failed to handle graph break gracefully ... BUILD_STRING type error`. The `Failed to handle graph break gracefully` wrapper double-counts the occurrence (38 total = 19 models × 2 reports per occurrence). Cited stack frame for the baseline model: `transformers/models/lfm2_vl/modeling_lfm2_vl.py:230-233` (the `torch_compilable_check(...)` call).

### Affected Models

| Model | Modes | Breaks |
|-------|-------|--------|
| AriaForConditionalGeneration | eval, train | 2 |
| AriaModel | eval, train | 2 |
| AyaVisionForConditionalGeneration | eval, train | 2 |
| Cohere2VisionForConditionalGeneration | eval, train | 2 |
| DeepseekVLForConditionalGeneration | eval, train | 2 |
| Gemma3ForConditionalGeneration | eval, train | 2 |
| GotOcr2ForConditionalGeneration | eval, train | 2 |
| GotOcr2Model | eval, train | 2 |
| InternVLForConditionalGeneration | eval, train | 2 |
| InternVLModel | eval, train | 2 |
| JanusForConditionalGeneration | eval, train | 2 |
| Lfm2VlForConditionalGeneration | eval, train | 2 |
| LightOnOcrForConditionalGeneration | eval, train | 2 |
| LlavaForConditionalGeneration | eval, train | 2 |
| Ovis2ForConditionalGeneration | eval, train | 2 |
| PaliGemmaForConditionalGeneration | eval, train | 2 |
| PaliGemmaModel | eval, train | 2 |
| Qwen2AudioForConditionalGeneration | eval, train | 2 |
| VipLlavaForConditionalGeneration | eval, train | 2 |

<details><summary>Verification signal (original)</summary>

`{"kind": "stderr_contains", "fragment": "BUILD_STRING type error"}`
</details>

## Pattern

The shared idiom across all 19 affected models. Cited call site at `transformers/models/lfm2_vl/modeling_lfm2_vl.py:230-233`:

```text
torch_compilable_check(
    inputs_embeds[special_image_mask].numel() == image_features.numel(),
    f"Image features and image tokens do not match, tokens: {n_image_tokens}, features: {n_image_features}",
)
```

`torch_compilable_check` (in `transformers.utils.import_utils`) wraps a non-callable `msg` argument into a closure. Inside the closure, the f-string is re-evaluated; with `dynamic=True`, `n_image_tokens` and `n_image_features` are SymNodes, so `BUILD_STRING` receives `SymNodeVariable` parts and rejects them (gb0021).

The two-stage nature of the break (data-dependent index `inputs_embeds[mask]` creates a resumption frame; closure-wrapped f-string then breaks on BUILD_STRING) is what makes it specific to torch's recent SymNode handling.

## Minimal reproducer (MRE)

Reduced from `modeling_lfm2_vl.py:230-233` (the cited `torch_compilable_check` call site). 35 lines (longer than ideal because the two-stage break requires preserving the data-dependent index that creates the resumption frame), runs in ~5s:

```python repro=true
"""
MRE for graph break: BUILD_STRING type error (gb0021).

Reduced from transformers' Lfm2VlForConditionalGeneration (and 18 other VL
models) get_placeholder_mask method
(transformers/models/lfm2_vl/modeling_lfm2_vl.py:230-233):

    torch_compilable_check(
        inputs_embeds[special_image_mask].numel() == image_features.numel(),
        f"... tokens: {n_image_tokens}, features: {n_image_features}",
    )

Trigger:
1. A data-dependent boolean-mask index (`inputs_embeds[mask]`) inside the
   `cond` argument forces a graph break (BINARY_SUBSCR on an unbacked dim).
2. Dynamo resumes compilation in a new frame after the break.
3. In the resumed frame, the f-string passed as `msg` interpolates
   `image_features.shape[0]`, which is a SymInt under dynamic=True.
4. transformers' `torch_compilable_check` (modeled here as `my_check`)
   wraps the str into a callable; inside the callable the BUILD_STRING
   bytecode receives a SymNodeVariable as one of its parts.
5. Dynamo's BUILD_STRING handler rejects SymNodeVariable parts:

    BUILD_STRING type error
      Explanation: Format string part type is not correct - expected
                   constant or format string.
      Developer debug context: SymNodeVariable()

Affects 19 transformers VL models (Lfm2Vl, Llava, PaliGemma, Aria, ...).

Run:
    python repro.py 2>&1 | grep "BUILD_STRING type error"
"""
import os

os.environ.setdefault("TORCH_LOGS", "graph_breaks")

import torch


def my_check(cond, msg):
    """Mirrors transformers.utils.import_utils.torch_compilable_check:
    wraps a non-callable msg into a closure, then calls torch._check."""
    if not callable(msg):
        _msg = msg
        def msg():
            return _msg
    torch._check(cond, msg)


@torch.compile(dynamic=True, backend="eager")
def fn(inputs_embeds, mask, image_features):
    n_tokens = mask.sum()
    n_features = image_features.shape[0]
    my_check(
        inputs_embeds[mask].numel() == image_features.numel(),
        f"tokens: {n_tokens}, features: {n_features}",
    )
    return inputs_embeds.sum()


inputs_embeds = torch.randn(4, 8)
mask = torch.tensor([[True] * 8, [False] * 8, [True] * 8, [False] * 8])
image_features = torch.randn(16, 1)
print(fn(inputs_embeds, mask, image_features))
```

<details><summary>Verification signal (MRE)</summary>

`{"kind": "stderr_contains", "fragment": "BUILD_STRING type error"}`
</details>

## Source

- Sweep results: `sweep_results/experiments/animesh-fullgraph-2026-04-28/explain_results.json` (rows for the 19 VL models above)
- Provenance anchor (baseline pick — fastest sweep wall-clock): `transformers/models/lfm2_vl/modeling_lfm2_vl.py:230-233`
- Built off PR pytorch/pytorch#181552
- Transformers: 5.6.2
- gb0021 doc: https://meta-pytorch.github.io/compile-graph-break-site/gb/gb0021.html

---
<sub>Body rewritten 2026-05-09 by file-issue + mre subagent dogfood. MRE deduced from real call site (not hypothesis), verified on torch 2.12 nightly. On torch 2.9 stable, MRE triggers different breaks — the BUILD_STRING path is torch-version-specific.</sub>
